File: central/app/services/retention.py
from __future__ import annotations

from app.storage.base import StorageBackend


def prune_old_snapshots(
    storage_backend: StorageBackend,
    namespace: str,
    keep_last: int,
) -> int:
    snapshots = storage_backend.list(namespace)
    if len(snapshots) <= keep_last:
        return 0

    snapshots.sort(key=lambda item: (item["mtime"], item["filename"]), reverse=True)
    to_delete = snapshots[keep_last:]

    for snapshot in to_delete:
        storage_backend.delete(namespace, snapshot["filename"])

    return len(to_delete)

File: central/internal/services/retention.go
package services

import (
	"cmp"
	"slices"

	"github.com/3to1go/central/internal/storage"
)

type snapshotStore interface {
	List(namespace string) ([]storage.StorageFile, error)
	Delete(namespace, filename string) error
}

// PruneOldSnapshots deletes excess snapshots, keeping the keepLast most recent.
// Returns the number of deleted snapshots.
func PruneOldSnapshots(backend snapshotStore, namespace string, keepLast int) (int, error) {
	files, err := backend.List(namespace)
	if err != nil {
		return 0, err
	}
	if len(files) <= keepLast {
		return 0, nil
	}

	slices.SortFunc(files, func(a, b storage.StorageFile) int {
		if a.Mtime != b.Mtime {
			return cmp.Compare(b.Mtime, a.Mtime)
		}
		return cmp.Compare(b.Filename, a.Filename)
	})

	toDelete := files[keepLast:]
	for _, f := range toDelete {
		if err := backend.Delete(namespace, f.Filename); err != nil {
			return 0, err
		}
	}
	return len(toDelete), nil
}
